src/igraph_eulerian.c

```diff
@@ -0,0 +1,565 @@
+*/
+int igraph_i_is_eulerian_undirected(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_integer_t odd;
+    igraph_vector_t degree, csize;
+    int i, cluster_count;
+
+    if (igraph_ecount(graph) == 0 || igraph_vcount(graph) <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return  IGRAPH_SUCCESS;
+    }
+
+    IGRAPH_CHECK(igraph_vector_init(&csize, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &csize);
+
+    IGRAPH_CHECK(igraph_clusters(graph, NULL, &csize, NULL, IGRAPH_WEAK));
+    cluster_count = 0;
+    for (i = 0; i < igraph_vector_size(&csize); i++) {
+        if (VECTOR(csize)[i] > 1) cluster_count++;
+    }
+
+    if (cluster_count > 1) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    odd = 0;
+
+    IGRAPH_CHECK(igraph_vector_init(&degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &degree);
+
+    IGRAPH_CHECK(igraph_degree(graph, &degree, igraph_vss_all(), IGRAPH_ALL, IGRAPH_LOOPS));
+
+    for (i = 0; i < igraph_vector_size(&degree); i++) {
+        if (((long int) VECTOR(degree)[i]) % 2 == 1) odd++;
+    }
+
+    igraph_vector_destroy(&csize);
+    igraph_vector_destroy(&degree);
+    IGRAPH_FINALLY_CLEAN(2);
+
+    if (odd > 2) {
+        *has_path = 0;
+        *has_cycle = 0;
+        return IGRAPH_SUCCESS;
+    }
+
+    if (odd == 2) {
+        *has_path = 1;
+        *has_cycle = 0;
+    } else {
+        *has_path = 1;
+        *has_cycle = 1;
+    }
+
+    return IGRAPH_SUCCESS;
+}
+
+
+int igraph_i_is_eulerian_directed(const igraph_t *graph, igraph_bool_t *has_path, igraph_bool_t *has_cycle) {
+    igraph_bool_t res_weak;
+    igraph_integer_t incoming_excess, outgoing_excess, clusters_weak;
+    int i, cluster_count;
+    igraph_vector_t out_degree, in_degree, csize_weak;
+
+    if (igraph_ecount(graph) == 0 || igraph_vcount(graph) <= 1) {
+        *has_path = 1;
+        *has_cycle = 1;
+        return IGRAPH_SUCCESS;
+    }
+
+    incoming_excess = 0;
+    outgoing_excess = 0;
+
+    res_weak = 1;
+
+    IGRAPH_CHECK(igraph_vector_init(&out_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &out_degree);
+    IGRAPH_CHECK(igraph_degree(graph, &out_degree, igraph_vss_all(), IGRAPH_OUT, IGRAPH_LOOPS));
+
+    IGRAPH_CHECK(igraph_vector_init(&in_degree, 0));
+    IGRAPH_FINALLY(igraph_vector_destroy, &in_degree);
+    IGRAPH_CHECK(igraph_degree(graph, &in_degree, igraph_vss_all(), IGRAPH_IN, IGRAPH_LOOPS));
+
+    /* checking if incoming vertices == outgoing vertices */
+    for (i = 0; i < igraph_vcount(graph); i++) {
+        if (VECTOR(in_degree)[i] != VECTOR(out_degree)[i]) {
+            if ((VECTOR(in_degree)[i] + 1 == VECTOR(out_degree)[i]) && (incoming_excess < 2 && outgoing_excess < 1)) {
+                outgoing_excess++;
+            } else if ((VECTOR(out_degree)[i] + 1 == VECTOR(in_degree)[i]) && (incoming_excess < 1 && outgoing_excess < 2)) {
```

Comment:
The reason why I did that is because outgoing_excess should be either 0 or 1, but we do not know which one it is, thats why I did that. The same applies to incoming_excess.

Reply:
I do not understand what you mean. As you say, `outgoing_excess` is either 0 or 1. The condition `outgoing_excess < 2` is always true. Why do you include the condition?

Reply:
fixed now